### ISSUE TITLE
Spacemacs support for Magit 2.1.0

### ISSUE THREAD

**Issue by syl20bnr** [magit/magit#1918]
I'm working on the support of magit 2.1 in Spacemacs and due to the high number of keymaps I added a new function to automatically adapt a keymap to evil (we call this `to evilify a map*`).
Good news is that it works great with *almost* all the keymaps in magit 2.1. I'm hitting a wall with the parent keymap `magit-mode-map` because it seems that `map-keymap` does not play well with `make-keymap` and `map-keymap` cannot iterate through the events.
If I change `make-keymap` to `make-sparse-keymap` then it works.

I have some questions regarding this:
1) What's the differences between `make-keymap` and `make-sparse-keymap` (besides the data structure) ? Does magit needs these differences ?
2) Do you have any advice to iterate trough the keymap defined with `make-keymap` ?
3) Ultimately, is it possible to use `make-sparse-keymap` ?

`*` *evilifying* a map is to introduce navigation, search and selection à la Vim. When an evilified key binding is shadowed then we move it according to the following rules (`k` -> `K`, `K` -> `C-k` and `C-k` ->  `C-S-k`). So an example in magit is `discard` is moved to `K` and `untrack` is moved to `C-k`, `log-popup` is moved to `L`, etc...

**Comment by npostavs** [comment]
1, 3) From the manual, "Format of Keymaps"
>    Several kinds of elements may appear in a keymap, after the symbol
‘keymap’ that begins it:
> ...
> ‘CHAR-TABLE’
     If an element of a keymap is a char-table, it counts as holding
     bindings for all character events with no modifier bits (*note
     modifier bits::): element N is the binding for the character with
     code N.  This is a compact way to record lots of bindings.  A
     keymap with such a char-table is called a "full keymap".  Other
     keymaps are called "sparse keymaps".

So basically it just saves space.

> 2) Do you have any advice to iterate trough the keymap defined with `make-keymap` ?

A char table acts like a vector so you can iterate over it like one. Or use `map-char-table`.

> we call this to `evilify a map`

Not [vilify](https://en.wiktionary.org/wiki/vilify)? I feel like you guys missed a great punning opportunity there.

**Comment by tarsius** [comment]
I was able to trace back the use of `make-keymap` for this keymap to 6edd6f3be774e91ebf65adb7bd24d8f8b1fd74e0 - which is before this package was even named Magit!

It does make a difference whether a full or sparse keymap is used. When a key is pressed which is self-inserting in the `current-global-map` and which is not overwritten in the effective (magit)  keymaps, then the use of a full map would result in an error about the buffer being read-only, while the use of a sparse keymap would result in an error about the key being unbound.

Not sure what's better. I'll have to think about it some more, maybe we can just use a sparse keymap.

> I'm working on the support of magit 2.1 in Spacemacs

I was gonna suggest you do that soon :-) I have some suggestions mostly ready. In what branch are you working on compatibility with magit 2.1?

> it seems that `map-keymap` does not play well with `make-keymap` and `map-keymap` cannot iterate through the events.

It has been a while since I played with that function. But I remember that I had more luck with `map-keymap-internal`. And at some point I wrote  [`kmu-map-keymap`](https://github.com/tarsius/keymap-utils/blob/master/keymap-utils.el#L521) and [`kmu-keymap-bindings`](https://github.com/tarsius/keymap-utils/blob/master/keymap-utils.el#L478`) - it seems I did so to deal with exactly this issue.

**Comment by tarsius** [comment]
When I looked at spacemacs' magit configuration a few days ago, I was wondering why the same changes were made to several keymaps which have `magit-mode-map` as parent instead of just in that map. Maybe that was necessary with releases before `2.1.0`, but it seems that starting with that release it would be enough to just change the bindings in `magit-mode-map`, and then a few select bindings in other maps where a binding from the shared map is overwritten in magit itself.

**Comment by syl20bnr** [comment]
> I was gonna suggest you do that soon :-) I have some suggestions mostly ready. In what branch are you working on compatibility with magit 2.1?

Indeed I should have looked at it sooner :-) Great you propose to look at it!

There is no special branch for the magit next support, you can checkout develop (you'll have to fetch the submodules to retrieve `magit-next`) then add the following layer to `dotspacemacs-configuration-layers` (you can go to it with `M-m f e h` and type `layers`):

```elisp
(git :variables git-use-magit-next t)
```

Then press `M-m f e R` to re-evaluate the configuration. Note that in Vim style `M-m` is `SPC` and that only `magit-next` is feched, the optional supported third parties by magit are not installed for now.

On current magit version 1.4 we are more or less applying a special `evil` state we call `evilified` and rely on `evil` API to handle the keymap override with `evil-define-key`, the `evilified state` is where we have our keymap for a minimalistic set of Vim bindings [1]. An example of such application can be found here [2].

[1] https://github.com/syl20bnr/spacemacs/blob/develop/spacemacs/extensions/evil-evilified-state/evil-evilified-state.el#L56-L69
[2] https://github.com/syl20bnr/spacemacs/blob/develop/contrib/!source-control/git/packages.el#L158-L166

For magit next, since there are more keymaps and the keymap text property is used, I experimented with automatic remapping [3] which applies the conventions defined in CONVENTIONS.org [4]. The configuration for magit next can be found here [5] (this is a WIP :-p). You can use `M-m f e h` to quickly go to the `magit-next` configuration.

[3] https://github.com/syl20bnr/spacemacs/blob/develop/core/core-evilify-keymap.el#L31-L35
[4] https://github.com/syl20bnr/spacemacs/blob/920f01a776b1f200c726dda6a7cc9cb602a99fd8/doc/CONVENTIONS.org#evilify-buffers
[5] https://github.com/syl20bnr/spacemacs/blob/develop/contrib/!source-control/git/extensions.el#L56-L82

I plan to release the version 103 at around the same time as magit 2.1. It should be possible I hope :-)

Thank you!!

**Comment by syl20bnr** [comment]
> Maybe that was necessary with releases before 2.1.0, but it seems that starting with that release it would be enough to just change the bindings in magit-mode-map, and then a few select bindings in other maps where a binding from the shared map is overwritten in magit itself.

I think we had to do it to make it work with the evil API. The new approach is even bigger since it processes all the maps, but maybe some of them are remained untouched so we could remove them. OTHO maybe we'll add some more keys to `evilified` state keymap so to be safe it is better to process all the magit keymaps.

Quick question, do you maintain a list of all the sections keymaps in the source code ? If so I could rely on it to process them.

**Comment by tarsius** [comment]
I have never used vim bindings. They have their appeal but I haven't found the time to dive in yet. But I also don't use the default emacs bindings, at least not for movement, so I know from first hand experience what cascade of remappings can be necessary, if all one wants to do is to change a select few. Where evil/vim use `hjkl`, I use `C-{jikl}`. But back to evil.

I don't think there even has to be a "evil-layer" in magit buffers, i.e. "emacs" and "vim" state could be the same, but I might be wrong. I think you could make just a few adjustments which are very similar to what I personally do, and be done with it:

```
default evil  command
p       j     magit-section-backward
n       k     magit-section-forward
M-p     J     magit-section-backward-sibling
M-n     K     magit-section-forward-sibling
k       n     magit-discard and others (memonic: *nu*uke change)
K       N     magit-file-untrack
v       p     magit-revert
V       P     magit-revert-popup (memonic: as *p*reviously)
```

By the way why can `v` not be used?

**Comment by tarsius** [comment]
Meeh... I see why that won't work.

**Comment by syl20bnr** [comment]
The evilified state has two purposes, first it is a minimalistic set of bindings so we don't shadow too much events and second it has its own color so we now that it is different than the regular Vim states.

For the key bindings we have to be very general because there are a lot of buffers to `evilify` so the two main ideas are minimalistic set of binding and dumb rules to rebind the bindings (by escalading them in complexity).

In the case of `v` and `V` they should be available on `C-v` and `C-S-v` respectively but if they are in `magit-mode-map` then it does not work for now with `make-keymap`.

**Comment by tarsius** [comment]
I am having a very hard time expressing my thoughts here, because I just know way to little about the vim side of things, and whatever I write ends up as a complicated mess and has to be discarded...

I understand that vim bindings in magit buffers should be consistent with vim bindings in other buffers. Still it feels wrong to just follow a simple remapping rule book.

By default all "important commands" in magit buffers are available from a alphabetic key - without any modifiers (or with `shift` as "modifier" for secondary variants). In a sense that is very vim-like I suppose. But some of those bindings conflict with conventional vim bindings. To make these essential vim bindings available, you make the other commands **less** vim-like using a binding which involves a modifier.

So Emacs users (who are used to `modifier+key`) users can just press a single key to invoke all the common commands (similar to Dired) but Vim users (who much prefer to press just `key`) cannot - that just seems ... strange.

I think there are two kinds of conflicts:

1. There are two essential "movement dimension": character/line vs. section. You have fixed that by using `C-j` and `C-k` for the section dimension. I don't know how hardcore vim users feel about that, but to me it seems fine.

2. Some of magit's mnemonic bindings conflict with essential vim bindings, and I don't feel the used remappings are good.

The problematic keys are `hjkl` and `v`.

* `h:help`: I think you have not remapped `magit-dispatch-popup` yet. Moving to `SPC something` seems fine to me. So no problem.

* `j:jump`:  I think you have not remapped jump commands yet. Moving to `SPC something` seems fine to me. So no problem.

* `l:log`, `k:discard`, and `v:revert`:

All of these are very important commands and I think the should be available from a strictly alphabetic binding. No modifiers.

I suggest that you move them to letters that are "free when using vim". Since in vim `n` and `p` are not movement commands, the *might* be free. (But they might be used for other essential (but non-movement) commands, I don't know). I would suggest using them for discard and revert.

I would very much prefer if all the "apply variants" (`magit-{apply,stage,unstage,discard,revert}`) were available from single character.

Logging is even more important. Moving to `L` *used to be* fine. Not good I would strongly favor a lowercase letter, but it's okay. But `L` is now bound to `magit-toggle-margin`, so you would have to find a new place for that.

**Comment by tarsius** [comment]
I think that functions that "automatically fix keymaps", such as `evilify` and `spacemacs/evilify-map` have to be able to deal with full keymaps (as created by `make-keymap`).

I might use `make-sparse-keymap` for `magit-mode-map` but lean toward not doing it. This shouldn't be a big deal for you, just do it manually for this particular map. But sooner or later you will run into other full keymaps. My `keymap-utils.el` library could serve as inspiration when implementing keymap related features that are missing from Emacs itself.

**Comment by tarsius** [comment]
A key that might work for `magit-log-popup` is `y`. `magit-cherry` which `Y` is bound to actually is a "log variant" so that would perfect sense.

The command bound to `y` by default is `magit-show-refs-popup`, we now have to move it elsewhere. (The old bindings made also sense because `magit-cherry` can also be considered a variant of `magit-show-refs-popup` (expanding a branch section in the refs buffer also shows cherries)).

`magit-show-refs-popup` could be moved to `T` which only makes a little sense (`t` creates tags, while this command lists - among other things - tags).

**Comment by tarsius** [comment]
Could `delete` and `backspace` be used for discard and revert?

**Comment by tarsius** [comment]
> For magit next, since there are more keymaps and the keymap text property is used, I experimented with automatic remapping

I have tried this approach myself in the past (I use `C-jikl` for navigation which is very problematic because "`C-i` is the same thing as `TAB`"). Several times. In the end I always ended up going back to doing it manually. Not saying you shouldn't do it - I haven't given up on the idea myself - just saying that automating the process might end up requiring a lot of work - so much that you might just as well have done it manually in the first place.

> Quick question, do you maintain a list of all the sections keymaps in the source code ?

No, you'll have to grep for `make-sparse-keymap`.

**Comment by tarsius** [comment]
s/revert/reverse/everywhere above

**Comment by tarsius** [comment]
I have decided to stick with `make-keymap`. `dired-mode-map` for example is a full keymap too, and the errors are just better.

**Comment by tarsius** [comment]
Could `.` and `,` be used?

**Comment by syl20bnr** [comment]
I finished the support for magit 2.1! :tada: And magit 2.1 is just amazing, thank you for such package :+1: 

Regarding the key bindings I opted for the automatic `evilification` (`vilifcation` for @npostavs ;-)) of maps via a macro, it has the advantage that once one knows the `evilified`keys (there are only 10 keys or so) then it is easy to look at the default magit key bindings in the magit manual and transpose them in evilified map. That is true that we get some kind of paradox between emacs users getting full single keys and vim users getting more keychords but I don't find it really annoying, in practice shifted keys works really well and getting `hjkl` navigation is more important.

As a side note we cannot alter easily the raw keymaps because Spacemacs must keep them untouched in emacs state (used by people not using the Vim key bindings) and the macro takes care of this burden now.

But once the initial pass of auto-evilification is done we can fine tune the keymaps if we can find keys that will not alter the stock keys set by magit (so people using Emacs key bindings are not bothered with it).

Once again, thank you for magit ! :+1:
